design/src/rfcs/rfc0047_custom_validation.md

```diff
@@ -0,0 +1,565 @@
+RFC: Custom Validation Exception
+================================
+
+> Status: RFC
+>
+> Applies to: server
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+This RFC defines a mechanism to use custom validation exception shapes instead of the standard `smithy.framework#ValidationException`, enabling service teams to maintain backward compatibility with existing APIs or use validation exceptions that have already been published to external consumers. This addresses scenarios where service teams are migrating from non-Smithy models or need to provide validation error responses in a format that differs from the standard Smithy validation exception structure.
+
+Terminology
+-----------
+
+- **Constrained shape**: a shape that is either:
+  1. a shape with a [constraint trait] attached
+  2. a (member) shape with a [`required` trait] attached  
+  3. an [`enum` shape]
+  4. an [`intEnum` shape]
+  5. a [`structure` shape] with at least one required member shape; or
+  6. a shape whose closure includes any of the above.
+- **ValidationException**: A Smithy error shape that is serialized in the response when constraint validation fails during request processing.
+- **Shape closure**: the set of shapes a shape can "reach", including itself.
+- **Custom validation exception**: A user-defined error shape marked with validation-specific traits that replaces the standard `smithy.framework#ValidationException`.
+
+[constraint trait]: https://smithy.io/2.0/spec/constraint-traits.html
+[`required` trait]: https://smithy.io/2.0/spec/type-refinement-traits.html#required-trait
+[`enum` shape]: https://smithy.io/2.0/spec/simple-types.html#enum
+[`intEnum` shape]: https://smithy.io/2.0/spec/simple-types.html#intenum
+[`structure` shape]: https://smithy.io/2.0/spec/aggregate-types.html#structure
+
+The user experience if this RFC is implemented
+----------------------------------------------
+
+Currently, if there is a constrained shape in the input shape closure of an operation, the operation **must** add `smithy.framework#ValidationException` to the operation's error list, otherwise a build error is raised:
+
+```
+Caused by: ValidationResult(shouldAbort=true, messages=[LogMessage(level=SEVERE, message=Operation com.aws.example#GetStorage takes in input that is constrained (https://smithy.io/2.0/spec/constraint-traits.html), and as such can fail with a validation exception. You must model this behavior in the operation shape in your model file.
```

Comment:
if this comes from smithy how are you disabling or getting around this validation? Also this seems like something we'd want to upstream to smithy anyway right?

Reply:
No, we do this as part of the server code here: https://github.com/smithy-lang/smithy-rs/blob/f18c70d36c40fa0f40860547394c134566704e69/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ValidateUnsupportedConstraints.kt#L224